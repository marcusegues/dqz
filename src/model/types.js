// @flow

export type CategoriesTypes =
  | 'Meat'
  | 'Butter'
  | 'Oils'
  | 'AlcSoft'
  | 'AlcHard'
  | 'Cigarettes'
  | 'Tobacco'
  | 'Meds'
  | 'Books'
  | 'Magazines'
  | 'Flowers'
  | 'AnimalFeed'
  | 'Fertilizer'
  | 'Other';

type CategoryInfo = {
  name: string,
  unit: string,
  icon: string,
};

export type CategoryInfoType = Map<CategoriesTypes, CategoryInfo>;

type CategoryVolume = {
  quantity: number,
  amounts: Array<number>,
};

type CategoryDues = {
  duty: number,
  vat: number,
};

type CategoryBasketItem = {
  volume: CategoryVolume,
  dues: CategoryDues,
};

export type BasketType = Map<CategoriesTypes, CategoryBasketItem>;

type DutyBracket = {
  threshold: number,
  fee: number,
};

type Rates = {
  vat: number,
  duty: Array<DutyBracket>,
  dutyAllowance: number,
  adultsOnly: boolean,
};

export type RatesType = Map<CategoriesTypes, Rates>;

export type People = {
  adults: number,
  minors: number,
};
