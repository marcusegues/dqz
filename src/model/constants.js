// @flow

import type {
  CategoriesTypes,
  CategoryInfoType,
  RatesType,
} from './types/basketPeopleTypes';
import Immutable from 'immutable';
import {
  makeCategoryInfoRecord,
  makeDutyBracketRecord,
  makeRatesRecord,
} from './types/basketPeopleTypes';

export const IndividualAllowanceAdult: number = 300;
export const IndividualAllowanceMinor: number = 300;

export const CATEGORIES: Set<CategoriesTypes> = new Set([
  'Meat',
  'Butter',
  'Oils',
  'OtherFood',
  'AlcSoft',
  'AlcHard',
  'Cigarettes',
  'Tobacco',
  'Meds',
  'Books',
  'Magazines',
  'Flowers',
  'AnimalFeed',
  'Fertilizer',
  'Other',
]);

export const CategoriesArray = Array.from(CATEGORIES.values());

export const CategoriesRates: RatesType = Immutable.Map([
  [
    'Meat',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List([
        makeDutyBracketRecord({ threshold: 10, fee: 17 }),
        makeDutyBracketRecord({ threshold: Infinity, fee: 23 }),
      ]),
      dutyAllowance: 1,
      adultsOnly: false,
    }),
  ],
  [
    'Butter',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List([
        makeDutyBracketRecord({ threshold: Infinity, fee: 16 }),
      ]),
      dutyAllowance: 1,
      adultsOnly: false,
    }),
  ],
  [
    'Oils',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List([
        makeDutyBracketRecord({ threshold: Infinity, fee: 2 }),
      ]),
      dutyAllowance: 5,
      adultsOnly: false,
    }),
  ],
  [
    'OtherFood',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List(),
      dutyAllowance: Infinity,
      adultsOnly: false,
    }),
  ],
  [
    'AlcSoft',
    makeRatesRecord({
      vat: 0.08,
      duty: Immutable.List([
        makeDutyBracketRecord({ threshold: Infinity, fee: 2 }),
      ]),
      dutyAllowance: 5,
      adultsOnly: true,
    }),
  ],
  [
    'AlcHard',
    makeRatesRecord({
      vat: 0.08,
      duty: Immutable.List([
        makeDutyBracketRecord({ threshold: Infinity, fee: 15 }),
      ]),
      dutyAllowance: 1,
      adultsOnly: true,
    }),
  ],
  [
    'Cigarettes',
    makeRatesRecord({
      vat: 0.08,
      duty: Immutable.List([
        makeDutyBracketRecord({ threshold: Infinity, fee: 0.25 }),
      ]),
      dutyAllowance: 250,
      adultsOnly: false,
    }),
  ],
  [
    'Tobacco',
    makeRatesRecord({
      vat: 0.08,
      duty: Immutable.List([
        makeDutyBracketRecord({ threshold: Infinity, fee: 0.1 }),
      ]),
      dutyAllowance: 250,
      adultsOnly: true,
    }),
  ],
  [
    'Meds',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List(),
      dutyAllowance: Infinity,
      adultsOnly: false,
    }),
  ],
  [
    'Books',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List(),
      dutyAllowance: Infinity,
      adultsOnly: false,
    }),
  ],
  [
    'Magazines',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List(),
      dutyAllowance: Infinity,
      adultsOnly: false,
    }),
  ],
  [
    'Flowers',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List(),
      dutyAllowance: Infinity,
      adultsOnly: false,
    }),
  ],
  [
    'AnimalFeed',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List(),
      dutyAllowance: Infinity,
      adultsOnly: false,
    }),
  ],
  [
    'Fertilizer',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List(),
      dutyAllowance: Infinity,
      adultsOnly: false,
    }),
  ],
  [
    'Other',
    makeRatesRecord({
      vat: 0.025,
      duty: Immutable.List(),
      dutyAllowance: Infinity,
      adultsOnly: false,
    }),
  ],
]);

export const CategoriesInfo: CategoryInfoType = Immutable.Map([
  [
    'Meat',
    makeCategoryInfoRecord({
      name: 'Fleisch und Fleischzubereitung',
      unit: 'kg',
      icon: 'meat',
    }),
  ],
  [
    'Butter',
    makeCategoryInfoRecord({
      name: 'Butter und Rahm',
      unit: 'kg/Liter',
      icon: 'butter',
    }),
  ],
  [
    'Oils',
    makeCategoryInfoRecord({
      name: 'Öle, Fette, Margarine zu Speisezwecken',
      unit: 'kg/Liter',
      icon: 'oil',
    }),
  ],
  [
    'OtherFood',
    makeCategoryInfoRecord({
      name: 'Sonstige Lebensmittel & Alkoholfreie Getranke',
      unit: 'kg/Liter',
      icon: 'oil',
    }),
  ],
  [
    'Other',
    makeCategoryInfoRecord({
      name: 'Sonstige Lebensmittel & Alkoholfreie Getranke',
      unit: 'kg/Liter',
      icon: 'oil',
    }),
  ],
  [
    'AlcSoft',
    makeCategoryInfoRecord({
      name: 'Alkoholische Getränke, Alkoholgehalt bis 18% Vol.',
      unit: 'Liter',
      icon: 'beer',
    }),
  ],
  [
    'AlcHard',
    makeCategoryInfoRecord({
      name: 'Alkoholische Getränke, Alkoholgehalt über 18% Vol.',
      unit: 'Liter',
      icon: 'vodka',
    }),
  ],
  [
    'Cigarettes',
    makeCategoryInfoRecord({
      name: 'Zigaretten/Zigarren',
      unit: 'Stück',
      icon: 'cigarette',
    }),
  ],
  [
    'Tobacco',
    makeCategoryInfoRecord({
      name: 'Andere Tabakfabrikate',
      unit: 'Gramm',
      icon: 'cigarette',
    }),
  ],
  [
    'Meds',
    makeCategoryInfoRecord({
      name: 'Medikamente',
      unit: '',
      icon: 'cigarette',
    }),
  ],
  [
    'Books',
    makeCategoryInfoRecord({
      name: 'Bücher',
      unit: '',
      icon: 'cigarette',
    }),
  ],
  [
    'Magazines',
    makeCategoryInfoRecord({
      name: 'Zeitschriften',
      unit: '',
      icon: 'cigarette',
    }),
  ],
  [
    'Flowers',
    makeCategoryInfoRecord({
      name: 'Schnittblumen & Pflanzen',
      unit: '',
      icon: 'cigarette',
    }),
  ],
  [
    'AnimalFeed',
    makeCategoryInfoRecord({
      name: 'Tierfutter',
      unit: '',
      icon: 'cigarette',
    }),
  ],
  [
    'Fertilizer',
    makeCategoryInfoRecord({
      name: 'Dünger',
      unit: '',
      icon: 'cigarette',
    }),
  ],
]);
