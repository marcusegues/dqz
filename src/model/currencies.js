// @flow
// $FlowFixMe
import { parseString } from 'react-native-xml2js';

export type Currency =
  | 'EUR'
  | 'USD'
  | 'CHF'
  | 'AUD'
  | 'CAD'
  | 'DKK'
  | 'GBP'
  | 'HKD'
  | 'JPY'
  | 'MXN'
  | 'NOK'
  | 'NZD'
  | 'RUB'
  | 'SEK'
  | 'ZAR';

export const currenciesArray: Array<Currency> = [
  'EUR',
  'USD',
  'CHF',
  'AUD',
  'CAD',
  'DKK',
  'GBP',
  'HKD',
  'JPY',
  'MXN',
  'NOK',
  'NZD',
  'RUB',
  'SEK',
  'ZAR',
];

export type CurrencyObject = { [Currency]: number };

export const currencyExample: CurrencyObject = {
  AUD: 0.77355,
  CAD: 0.78784,
  CHF: 1,
  DKK: 0.15924670000000002,
  EUR: 1.17796, // fixed for Maerki Example
  GBP: 1.33658,
  HKD: 0.1279084,
  JPY: 0.0088304,
  MXN: 0.05034,
  NOK: 0.1203345,
  NZD: 0.70344,
  RUB: 0.017327699999999998,
  SEK: 0.1201607,
  USD: 0.99205, // fixed for Maerki Example
  ZAR: 0.07995,
};

// export const getFilteredCurrencies = (): CurrencyObject => {};

export const parseCurrencyXML = (rawdata: any, store: any): void => {
  let currencyObject: CurrencyObject = { CHF: 1 };
  let validCurrencies: boolean = false;
  let currencyDate: Date = new Date('2000-01-01');

  parseString(rawdata, (e, r) => {
    if (!r) {
      // console.log('Error Case');
      currencyObject = currencyExample;
    } else {
      // console.log('Normal Case');
      validCurrencies = true;
      const { wechselkurse } = r;
      const { devise, datum } = wechselkurse;
      currencyDate = new Date(datum[0]);
      devise
        .filter(d => currenciesArray.indexOf(d.$.code.toUpperCase()) > -1)
        .forEach(c => {
          const kurs = c.kurs[0];
          const waehrung = c.waehrung[0];
          const multiplier = +waehrung.replace(/ .*/, '');
          currencyObject[c.$.code.toUpperCase()] = kurs / multiplier;
        });
    }
  });
  store.dispatch({
    type: 'UPDATE_CURRENCIES',
    currencyObject,
    validCurrencies,
    currencyDate,
  });
};
