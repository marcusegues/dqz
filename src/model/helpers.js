// @flow

import Immutable from 'immutable';
import type { Set as ImmutableSetType } from 'immutable';
import type {
  BasketType,
  CategoriesTypes,
  CategoryBasketItem,
  People,
} from './types/types';
import { CategoriesArray } from './constants';
import { makeCategoryBasketItemRecord, makePeopleRecord } from './types/types';

const emptyItem: CategoryBasketItem = makeCategoryBasketItemRecord();

export const emptyBasket: BasketType = Immutable.Map(
  CategoriesArray.map(c => [c, emptyItem])
);

// QUANTITIES
export const setQuantity = (
  basket: BasketType,
  category: CategoriesTypes,
  quantity: number
): BasketType => basket.setIn([category, 'volume', 'quantity'], quantity);

export const setQuantities = (
  basket: BasketType,
  pairs: Array<{ category: CategoriesTypes, quantity: number }>
): BasketType => {
  return basket.withMutations(bskt => {
    pairs.forEach(p => setQuantity(bskt, p.category, p.quantity));
  });
};

export const getQuantity = (
  basket: BasketType,
  category: CategoriesTypes
): number => Immutable.getIn(basket, [category, 'volume', 'quantity'], 0);

export const addAmount = (
  basket: BasketType,
  category: CategoriesTypes,
  amount: number
): BasketType =>
  basket.updateIn([category, 'volume', 'amounts'], Immutable.Set(), amounts =>
    amounts.add(amount)
  );

export const resetAmounts = (
  basket: BasketType,
  category: CategoriesTypes
): BasketType => basket.setIn([category, 'volume', 'amounts'], Immutable.Set());

export const getAmounts = (
  basket: BasketType,
  category: CategoriesTypes
): ImmutableSetType<number> =>
  Immutable.getIn(basket, [category, 'volume', 'amounts'], Immutable.Set());

export const initPeople: People = makePeopleRecord({
  adults: 1,
  minors: 0,
});
