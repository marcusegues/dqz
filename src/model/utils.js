// @flow
import type { Amounts } from './types/basketPeopleAmountsTypes';
import type { Currency, CurrencyObject } from './currencies';
import { INDIVIDUALALLOWANCE } from './constants';

export const rounding = (x: number): number => {
  // this rounding is not perfect, due to floating point
  const rounding1: number = Math.round(x * 100);
  const modulo: number = rounding1 % 5;
  let adjusted = rounding1;
  switch (modulo) {
    case 1:
      adjusted -= 1;
      break;
    case 2:
      adjusted -= 2;
      break;
    case 3:
      adjusted += 2;
      break;
    case 4:
      adjusted += 1;
      break;
    default:
  }

  return adjusted / 100;
};

export const formatDate = (d: Date): string => {
  const mm = d.getMonth() + 1;
  const dd = d.getDate();
  // eslint-disable-next-line no-restricted-globals
  if (isNaN(mm) || isNaN(dd)) {
    return 'Invalid Date';
  }
  return [
    (dd > 9 ? '' : '0') + dd,
    (mm > 9 ? '' : '0') + mm,
    d.getFullYear(),
  ].join('.');
};

export const formatFullDate = (d: string, hours: string, minutes: string) => `${d}/${hours}:${minutes}`;

export type FlatAmount = {
  currency: Currency,
  amount: number,
  large: boolean,
  id: string,
};

export const flatAmounts = (amounts: Amounts): Array<FlatAmount> => {
  const result: Array<FlatAmount> = [];
  amounts.forEach((v, k) => {
    v.amounts.map(amt =>
      result.push({ currency: k, amount: amt.amount, large: false, id: amt.id })
    );
    v.largeAmounts.map(amt =>
      result.push({ currency: k, amount: amt.amount, large: true, id: amt.id })
    );
  });
  return result;
};

export const flatAllAmounts = (amounts: Amounts): Array<FlatAmount> =>
  flatAmounts(amounts).filter(a => !a.large);

export const flatLargeAmounts = (amounts: Amounts): Array<FlatAmount> =>
  flatAmounts(amounts).filter(a => a.large);

export const hasLargeAmount = (
  amounts: Amounts,
  currencyObject: CurrencyObject
): boolean =>
  flatAllAmounts(amounts).reduce(
    (a, v) => Math.max(a, v.amount * currencyObject[v.currency]),
    0
  ) > INDIVIDUALALLOWANCE;

export const hasOffsettingAmount = (
  amounts: Amounts,
  amount: number,
  currency: Currency
): boolean => {
  const totalsOfCurrency = flatAllAmounts(amounts)
    .filter(a => a.currency === currency)
    .reduce((a, v) => v.amount + a, 0);
  const largePerCurrency = flatLargeAmounts(amounts)
    .filter(a => a.currency === currency)
    .reduce((a, v) => v.amount + a, 0);
  return totalsOfCurrency - largePerCurrency > amount;
};

export const totalAllAmounts = (
  amounts: Amounts,
  currencyObject: CurrencyObject
): number => {
  const flat = flatAmounts(amounts);
  return flat.reduce((a, v) => a + v.amount * currencyObject[v.currency], 0);
};
