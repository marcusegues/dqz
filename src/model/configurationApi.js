// @flow

import Immutable from 'immutable';
import { v4 } from 'uuid';
import type { List as ImmutableListType } from 'immutable';
import type {
  Amounts,
  AmountWithId,
  Basket,
  Category,
  CategoryBasketItem,
  People,
  Quantity,
} from './types/basketPeopleAmountsTypes';
import { categoriesArray } from './constants';
import {
  makeAmountsOfCurrencyRecord,
  makeCategoryBasketItemRecord,
  makePeopleRecord,
} from './types/basketPeopleAmountsTypes';
import type { Currency } from './currencies';
import { flatAllAmounts, getConvertedLocalTimeToSwiss } from './utils';
import type { FlatAmount } from './utils';

const emptyItem: CategoryBasketItem = makeCategoryBasketItemRecord();

/**
 * Empty Basket
 * @type Basket
 */
export const emptyBasket: Basket = Immutable.Map(
  categoriesArray.map(c => [c, emptyItem])
);

// QUANTITIES
/**
 * Adds a quantity in a basket for a given category
 * @param basket
 * @param category
 * @param quantity
 * @returns Basket
 */
export const addQuantity = (
  basket: Basket,
  category: Category,
  quantity: number
): Basket => {
  if (quantity <= 0) {
    return basket;
  }
  return basket.updateIn([category, 'volume', 'quantities'], q =>
    q.push({
      number: quantity,
      date: getConvertedLocalTimeToSwiss().toString(),
    })
  );
};

/**
 * Deletes a quantity in a basket for a given category at index
 * Note: if index > size of quantities list, quantities won't change.
 * This exception should be handled elsewhere.
 * @param basket
 * @param category
 * @param index
 * @returns Basket
 */
export const deleteQuantity = (
  basket: Basket,
  category: Category,
  index: number
): Basket =>
  basket.updateIn([category, 'volume', 'quantities'], q => q.delete(index));

/**
 * Reset all quantites in a basket for a given category.
 * @param basket
 * @param category
 * @returns Basket
 */
export const resetQuantities = (basket: Basket, category: Category): Basket =>
  basket.setIn([category, 'volume', 'quantities'], Immutable.List());

/**
 * Reset all quantites in a basket for a given category.
 * @param basket
 * @param categories (array of category)
 * @returns Basket
 */
export const resetQuantitiesMultipleCategories = (
  basket: Basket,
  categories: Array<Category>
): Basket =>
  basket.withMutations(bskt => {
    categories.forEach(category => {
      bskt.setIn([category, 'volume', 'quantities'], Immutable.List());
    });
    return bskt;
  });

/**
 * Helper method, not to be used in the App.
 * Sets several quantites for the given basket in different categories
 * @param basket
 * @param pairs
 * @returns Basket
 */
export const addQuantities = (
  basket: Basket,
  pairs: Array<{ category: Category, quantity: number }>
): Basket =>
  basket.withMutations(bskt => {
    pairs.forEach(p => addQuantity(bskt, p.category, p.quantity));
  });

/**
 * Gets all quantities for the basket's category
 * @param basket
 * @param category
 * @returns List of quantities (object)
 */
export const getQuantities = (
  basket: Basket,
  category: Category
): ImmutableListType<Quantity> =>
  basket.getIn([category, 'volume', 'quantities'], Immutable.List());

/**
 * Helper to get the quantity number from a quantity
 * @param quantity
 * @returns {number}
 */
export const getQuantityNumber = (quantity: Quantity): number =>
  quantity.number || 0;

/**
 * Helper to get the quantity date (string) from a quantity
 * @param quantity
 * @returns {*}
 */
export const getQuantityDate = (quantity: Quantity): string => quantity.date;

/**
 * Gets the total quantity for the basket's category
 * @param basket
 * @param category
 * @returns number
 */
export const getTotalQuantity = (
  basket: Basket,
  category: Category
): number => basket
    .getIn([category, 'volume', 'quantities'], Immutable.List())
    .reduce((a, v) => a + getQuantityNumber(v), 0);

// AMOUNTS
export const initAmounts: Amounts = Immutable.Map();

/**
 * Helper to create currency section if not present
 */
const addCurrencyIfNeeded = (amounts: Amounts, currency: Currency): Amounts => {
  if (amounts.has(currency)) {
    return amounts;
  }
  return amounts.set(currency, makeAmountsOfCurrencyRecord());
};

/**
 * Adds an amount in a given category of a basket
 * @param amounts
 * @param currency
 * @param amount
 * @returns Amounts type
 */
export const addAmount = (
  amounts: Amounts,
  currency: Currency,
  amount: number
): Amounts =>
  addCurrencyIfNeeded(amounts, currency).updateIn(
    [currency, 'amounts'],
    Immutable.List(),
    list => list.push({ amount, id: v4() })
  );

/**
 * Resets the amounts in a basket's category
 * @param amounts
 * @param currency
 * @returns Amounts type
 */
export const resetAmounts = (amounts: Amounts, currency: Currency): Amounts =>
  addCurrencyIfNeeded(amounts, currency).setIn(
    [currency, 'amounts'],
    Immutable.List()
  );

/**
 * Deletes amount with a given ID
 * @param amounts
 * @param id
 * @returns {Amounts}
 */
export const deleteAmount = (amounts: Amounts, id: string): Amounts => {
  const flat = flatAllAmounts(amounts);
  const element: ?FlatAmount = flat.find(a => a.id === id);
  if (!element) {
    return amounts;
  }
  const amountsKey = element.large ? 'largeAmounts' : 'amounts';
  const amts = amounts.getIn([element.currency, amountsKey], Immutable.List());
  const listIdx = amts.findIndex(v => v.id === id);
  const newAmounts = amts.delete(listIdx);
  return amounts.setIn([element.currency, amountsKey], newAmounts);
};

/**
 * Gets amount at index
 * @param amounts
 * @param id
 * @returns The amount (type FlatAmount) or undefined
 */
export const getAmount = (amounts: Amounts, id: string): FlatAmount => {
  const flat = flatAllAmounts(amounts);
  const el = flat.find(a => a.id === id);
  if (el) {
    return el;
  }
  return { amount: 0, currency: 'CHF', large: false, id: 'N/A' };
};

/**
 * Gets the amounts of a basket's category
 * @param amounts
 * @param currency
 * @returns ImmutableListType<number>
 */
export const getAmounts = (
  amounts: Amounts,
  currency: Currency
): ImmutableListType<AmountWithId> =>
  amounts.getIn([currency, 'amounts'], Immutable.List());

// AMOUNTS LARGE
/**
 * Adds a large amount in a given category of a basket
 * @param amounts
 * @param currency
 * @param amountLarge
 * @returns Amount type
 */
export const addLargeAmount = (
  amounts: Amounts,
  currency: Currency,
  amountLarge: number
): Amounts =>
  addCurrencyIfNeeded(amounts, currency).updateIn(
    [currency, 'largeAmounts'],
    Immutable.List(),
    list => list.push({ amount: amountLarge, id: v4() })
  );

/**
 * Resets the large amounts in a basket's category
 * @param amounts
 * @param currency
 * @returns Basket
 */
export const resetLargeAmounts = (
  amounts: Amounts,
  currency: Currency
): Amounts =>
  addCurrencyIfNeeded(amounts, currency).setIn(
    [currency, 'largeAmounts'],
    Immutable.List()
  );

/**
 * Gets the large amounts of a basket's category
 * @param amounts
 * @param currency
 * @returns ImmutableListType<AmountWithId>
 */
export const getLargeAmounts = (
  amounts: Amounts,
  currency: Currency
): ImmutableListType<AmountWithId> =>
  amounts.getIn([currency, 'largeAmounts'], Immutable.List());

/**
 * Returns an init people configuration (1 adult, no minor)
 */
export const initPeople: People = makePeopleRecord({
  adults: 1,
  minors: 0,
});

/**
 * Gets the people in the party (adult + minor)
 * @param party
 * @returns number
 */
export const getTotalPeople = (party: People): number =>
  party.get('adults') + party.get('minors');

/**
 * Gets the adult people in the party
 * @param party
 * @returns number
 */
export const getAdultPeople = (party: People): number => party.get('adults');

/**
 * Gets the minor people in the party
 * @param party
 * @returns number
 */
export const getMinorPeople = (party: People): number => party.get('minors');

/**
 * Adds an adult to the party
 * @param party
 * @returns People Type
 */
export const addAdult = (party: People): People =>
  party.update('adults', a => a + 1);

/**
 * Sets the number of adults in a party (non-neg)
 * @param party
 * @param n
 * @returns People Type
 */
export const setAdultPeople = (party: People, n: number): People =>
  party.set('adults', Math.max(0, n));

/**
 * Subtracts an adult from the party (non-negative result)
 * @param party
 * @returns People Type
 */
export const subtractAdult = (party: People): People =>
  party.update('adults', a => Math.max(a - 1, 0));

/**
 * Adds a minor to the party
 * @param party
 * @returns People Type
 */
export const addMinor = (party: People): People =>
  party.update('minors', m => m + 1);

/**
 * Sets the number of adults in a party (non-neg)
 * @param party
 * @param n
 * @returns People Type
 */
export const setMinorPeople = (party: People, n: number): People =>
  party.set('minors', Math.max(0, n));

/**
 * Subtracts a minor from the party (non-negative result)
 * @param party
 * @returns People Type
 */
export const subtractMinor = (party: People): People =>
  party.update('minors', m => Math.max(m - 1, 0));

/**
 * Set all people directly (non-negative)
 * @param party
 * @param adults
 * @param minors
 * @returns People Type
 */
export const setPeople = (
  party: People,
  adults: number,
  minors: number
): People => {
  const p1 = party.set('minors', Math.max(0, minors));
  return p1.set('adults', Math.max(0, adults));
};
