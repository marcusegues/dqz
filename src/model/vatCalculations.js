// @flow
import Immutable from 'immutable';
import type { Amounts, People } from './types/basketPeopleAmountsTypes';
import { INDIVIDUALALLOWANCE, OVERALLVATRATE } from './constants';
import { makeVatReportRecord } from './types/calculationTypes';
import type { Currency, CurrencyObject } from './currencies';
import { currenciesArray } from './currencies';
import { makeAmountsOfCurrencyRecord } from './types/basketPeopleAmountsTypes';
import type { VatReport } from './types/calculationTypes';
import { rounding } from './utils';

type totalPerCurrency = {
  amounts: number,
  amountsLarge: number,
};

const getCHFperCurrency = (
  amounts: Amounts,
  currency: Currency,
  currencyObject: CurrencyObject
): totalPerCurrency => {
  const thisCurrency = amounts.get(currency, makeAmountsOfCurrencyRecord());
  const sum = (a: number, v: number): number => a + v;

  const amountsThisCurrency: number = thisCurrency
    .get('amounts', Immutable.List())
    .map(a => a.amount)
    .reduce(sum, 0);
  const amountsLargeThisCurrency: number = thisCurrency
    .get('amountsLarge', Immutable.List())
    .map(a => a.amount)
    .reduce(sum, 0);

  const exchangeRate: number = currencyObject[currency] || 1;

  return {
    amounts: Math.floor(exchangeRate * (amountsThisCurrency - amountsLargeThisCurrency)),
    amountsLarge: Math.floor(exchangeRate * amountsLargeThisCurrency),
  };
};

export const calculateAllowancesExceptLast = (people: People): number =>
  (people.get('minors') + people.get('adults') - 1) * INDIVIDUALALLOWANCE;

export const calculateVat = (
  amounts: Amounts,
  people: People,
  currencyObject: CurrencyObject
): VatReport => {
  const sumsPerCurrency: Array<totalPerCurrency> = currenciesArray.map(c =>
    getCHFperCurrency(amounts, c, currencyObject)
  );

  const grandTotals: totalPerCurrency = sumsPerCurrency.reduce(
    (a, v) => {
      a.amounts += v.amounts;
      a.amountsLarge += v.amountsLarge;
      return a;
    },
    { amounts: 0, amountsLarge: 0 }
  );

  const afterAllowance = Math.max(
    0,
    grandTotals.amounts - calculateAllowancesExceptLast(people)
  );
  const effectiveAllowance: number = grandTotals.amounts - afterAllowance;

  grandTotals.amounts = afterAllowance;

  const sumLastPerson = grandTotals.amounts + grandTotals.amountsLarge;

  if (sumLastPerson <= INDIVIDUALALLOWANCE) {
    return makeVatReportRecord({
      totalVat: 0,
    });
  }

  const lastStep: number =
    rounding(OVERALLVATRATE * grandTotals.amounts) +
    rounding(OVERALLVATRATE * grandTotals.amountsLarge);

  return makeVatReportRecord({
    totalVat: lastStep,
    totalAmountsApprox: (
      grandTotals.amounts +
      grandTotals.amountsLarge +
      effectiveAllowance
    ).toFixed(2),
  });
};
