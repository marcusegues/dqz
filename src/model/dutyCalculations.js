// @flow

import type {
  Map as ImmutableMapType,
  List as ImmutableListType
} from 'immutable';
import Immutable from 'immutable';
import type {
  Basket,
  Category,
  DutyBracket,
  People
} from './types/basketPeopleTypes';
import type { DutyReport } from './types/calculationTypes';
import { CategoriesArray, CategoriesRates } from './constants';
import { makeDutyReportRecord } from './types/calculationTypes';
import { rounding } from './utils';
import { getQuantity } from './configurationApi';

export const calculateDuty = (basket: Basket, people: People): DutyReport => {
  let total = 0;
  const reportByCategory: ImmutableMapType<
    Category,
    number
  > = Immutable.Map().withMutations(r => {
    CategoriesArray.forEach(c => {
      const quantityRaw: number = getQuantity(basket, c);
      const adultsOnly: boolean = CategoriesRates.getIn(
        [c, 'adultsOnly'],
        false
      );
      const peopleCount: number =
        people.get('adults', 0) + +!adultsOnly * people.get('minors', 0);

      let allowanceRaw: number = CategoriesRates.getIn([c, 'dutyAllowance'], 0);
      const dutyDependency: ?Category = CategoriesRates.getIn(
        [c, 'dutyAllowanceDependency'],
        null
      );
      allowanceRaw *= peopleCount;
      if (dutyDependency) {
        allowanceRaw -= getQuantity(basket, dutyDependency);
        allowanceRaw = Math.max(0, allowanceRaw);
      }

      const quantity: number = quantityRaw - allowanceRaw;

      let allowanceRunningTotal: number = 0;
      const duty: ImmutableListType<DutyBracket> = CategoriesRates.getIn(
        [c, 'duty'],
        Immutable.List()
      );

      if (quantity <= allowanceRunningTotal) {
        r.set(c, 0);
      } else {
        const fee = duty.reduce((acc, v) => {
          const tempQuantity = Math.max(0, quantity - allowanceRunningTotal);
          let thresholdThisBracket: number = v.get('threshold', 0);
          if (thresholdThisBracket !== Infinity) {
            thresholdThisBracket = peopleCount * v.get('threshold', 0);
          }
          const bracketWidth = thresholdThisBracket - allowanceRunningTotal;
          const inThisBracket = Math.min(bracketWidth, tempQuantity);
          allowanceRunningTotal += bracketWidth;
          return acc + rounding(inThisBracket * v.get('fee'));
        }, 0);
        r.set(c, fee);
        total += fee;
      }
    });
  });
  return makeDutyReportRecord({
    totalDuty: total,
    dutyByCategoryRaw: reportByCategory
  });
};
