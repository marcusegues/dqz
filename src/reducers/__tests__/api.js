// @flow
/* global describe, test, expect, jest */
import Immutable from 'immutable';
import {
  getInitialState,
  makeSettingsRecord,
} from '../../types/reducers/appReducer';
import type { State } from '../../types/reducers/appReducer';
import { emptyBasket, initPeople } from '../../model/configurationApi';
import {
  getBasket,
  getPeople,
  getAmounts,
  getMainCategories,
  getSettings,
  getCurrencies,
  getCurrencyState,
  getTotalVat,
  getTotalDuty,
  getTotalFees,
  getVatReport,
  getDutyReport,
} from '../index';
import {
  sampleAmounts1,
  sampleBasket1,
} from '../../model/__tests__/fullBasketsAndAmounts';

jest.mock('uuid', () => ({
  v4: jest.fn(() => 1),
}));

const initState: State = getInitialState();

const amountAndQuantityState = initState
  .set('amounts', sampleAmounts1)
  .set('basket', sampleBasket1);

describe('API', () => {
  test('getBasket', () => {
    expect(getBasket({ appState: initState }).toString()).toBe(
      emptyBasket.toString()
    );
  });
  test('getPeople', () => {
    expect(getPeople({ appState: initState }).toString()).toBe(
      initPeople.toString()
    );
  });
  test('getAmounts', () => {
    expect(getAmounts({ appState: initState }).toString()).toBe(
      Immutable.Map().toString()
    );
  });
  test('getMainCategories', () => {
    expect(getMainCategories({ appState: initState }).toString()).toBe(
      Immutable.Set().toString()
    );
  });
  test('getSettings', () => {
    expect(getSettings({ appState: initState }).toString()).toBe(
      makeSettingsRecord().toString()
    );
  });
  test('getCurrencies', () => {
    expect(getCurrencies({ appState: initState }).toString()).toBe(
      {}.toString()
    );
  });
  test('getCurrencyState', () => {
    expect(getCurrencyState({ appState: initState }).toString()).toBe(
      false.toString()
    );
  });
  test('getTotalVat', () => {
    expect(getTotalVat({ appState: amountAndQuantityState })).toBe(485.4);
  });
  test('getTotalDuty', () => {
    expect(getTotalDuty({ appState: amountAndQuantityState })).toBe(3118.9);
  });
  test('getTotalFees', () => {
    expect(getTotalFees({ appState: amountAndQuantityState })).toBe(3604.3);
  });
  test('getVatReport', () => {
    expect(
      getVatReport({ appState: amountAndQuantityState })
    ).toMatchSnapshot();
  });
  test('getDutyReport', () => {
    expect(
      getDutyReport({ appState: amountAndQuantityState })
    ).toMatchSnapshot();
  });
});
