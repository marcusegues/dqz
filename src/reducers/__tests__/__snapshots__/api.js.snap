// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API getDutyReport 1`] = `
Immutable.Record {
  "totalDuty": 3118.9,
  "dutyByCategoryRaw": Immutable.Map {
    "Butter": 352,
    "Oils": 0,
    "Cigarettes": 0,
    "AlcSoft": 0,
    "Fertilizer": 0,
    "AlcHard": 0,
    "Meds": 0,
    "Books": 0,
    "Meat": 2470,
    "OtherFood": 0,
    "Other": 0,
    "Tobacco": 296.9,
    "Flowers": 0,
    "AnimalFeed": 0,
    "Magazines": 0,
  },
}
`;

exports[`API getVatReport 1`] = `
Immutable.Record {
  "totalVat": 485.4,
  "totalAmountsApprox": "6304.00",
}
`;
